metasrc/tex/lsstbib.py

```diff
@@ -0,0 +1,270 @@
+
+        .. code-block:: python
+
+           ['lsst', 'lsst-dm', 'refs', 'books', 'refs_ads']
+
+    Returns
+    -------
+    bibtex : `dict`
+        Dictionary with keys that are bibtex file names (such as ``'lsst'``,
+        ``'lsst-dm'``). Values are the corresponding bibtex file content
+        (`str`).
+    """
+    logger = logging.getLogger(__name__)
+
+    if bibtex_filenames is None:
+        # Default lsst-texmf bibliography files
+        bibtex_names = KNOWN_LSSTTEXMF_BIB_NAMES
+    else:
+        # Sanitize filenames (remove extensions, path)
+        bibtex_names = []
+        for filename in bibtex_filenames:
+            name = os.path.basename(os.path.splitext(filename)[0])
+            if name not in KNOWN_LSSTTEXMF_BIB_NAMES:
+                logger.warning('%r is not a known lsst-texmf bib file',
+                               name)
+                continue
+            bibtex_names.append(name)
+
+    # names of bibtex files not in cache
+    uncached_names = [name for name in bibtex_names
+                      if name not in _LSSTTEXMF_BIB_CACHE]
+    if len(uncached_names) > 0:
+        # Download bibtex and put into the cache
+        loop = asyncio.get_event_loop()
+        future = asyncio.ensure_future(_download_lsst_bibtex(uncached_names))
+        loop.run_until_complete(future)
+        for name, text in zip(bibtex_names, future.result()):
+            _LSSTTEXMF_BIB_CACHE[name] = text
+
+    return {name: _LSSTTEXMF_BIB_CACHE[name] for name in bibtex_names}
+
+
+def get_bibliography(lsst_bib_names=None, bibtex=None):
+    """Make a pybtex BibliographyData instance from standard lsst-texmf
+    bibliography files and user-supplied bibtex content.
+
+    Parameters
+    ----------
+    lsst_bib_names : sequence of `str`, optional
+        Names of lsst-texmf BibTeX files to include. For example:
+
+        .. code-block:: python
+
+           ['lsst', 'lsst-dm', 'refs', 'books', 'refs_ads']
+
+        Default is `None`, which includes all lsst-texmf bibtex files.
+
+    bibtex : `str`
+        BibTeX source content not included in lsst-texmf. This can be content
+        from a import ``local.bib`` file.
+
+    Returns
+    -------
+    bibliography : `pybtex.database.BibliographyData`
+        A pybtex bibliography database that includes all given sources:
+        lsst-texmf bibliographies and ``bibtex``.
+    """
+    bibtex_data = get_lsst_bibtex(bibtex_filenames=lsst_bib_names)
+
+    # Parse with pybtex into BibliographyData instances
+    pybtex_data = [pybtex.database.parse_string(_bibtex, 'bibtex')
+                   for _bibtex in bibtex_data.values()]
+
+    # Also parse local bibtex content
+    if bibtex is not None:
+        pybtex_data.append(pybtex.database.parse_string(bibtex, 'bibtex'))
+
+    # Merge BibliographyData
+    bib = pybtex_data[0]
+    if len(pybtex_data) > 1:
+        for other_bib in pybtex_data[1:]:
+            for key, entry in other_bib.entries.items():
```

Comment:
`bib.add_entries(other_bib.entries)` ?

Reply:
I tried that, but surprisingly it's broken in pybtex. I should upstream a fix.